pkg/imgpkg/imageset/image_set.go

```diff
@@ -77,57 +79,135 @@ func (o *ImageSet) Import(imgOrIndexes []imagedesc.ImageOrIndex,
 	defer func() { o.logger.WriteStr("imported %d images\n", len(importedImages.All())) }()
 
-	errCh := make(chan error, len(imgOrIndexes))
 	importThrottle := util.NewThrottle(o.concurrency)
 
+	imageOrIndexesToWrite := map[regname.Reference]regremote.Taggable{}
+	var imageOrIndexesToWriteLock = &sync.Mutex{}
+	errCh := make(chan error, len(imgOrIndexes))
 	for _, item := range imgOrIndexes {
 		item := item // copy
 
 		go func() {
 			importThrottle.Take()
 			defer importThrottle.Done()
+			addImageOrImageIndexToTaggableMapForMultiWrite(item, errCh, importRepo, registry, imageOrIndexesToWrite, imageOrIndexesToWriteLock)
+		}()
+	}
 
-			existingRef, err := regname.NewDigest(item.Ref())
-			if err != nil {
-				errCh <- err
-				return
-			}
+	err := checkForAnyAsyncErrors(imgOrIndexes, errCh)
+	if err != nil {
+		return nil, err
+	}
 
-			importDigestRef, err := o.importImage(item, existingRef, importRepo, registry)
-			if err != nil {
-				errCh <- fmt.Errorf("Importing image %s: %s", existingRef.Name(), err)
-				return
-			}
+	err = registry.MultiWrite(imageOrIndexesToWrite, o.concurrency)
+	if err != nil {
+		return nil, err
+	}
 
-			var regImage regv1.Image
-			if item.Image != nil {
-				regImage = *item.Image
-			}
-			var regImageIndex regv1.ImageIndex
-			if item.Index != nil {
-				regImageIndex = *item.Index
-			}
-			importedImages.Add(ProcessedImage{
-				UnprocessedImageRef: UnprocessedImageRef{existingRef.Name(), item.Tag()},
-				DigestRef:           importDigestRef.Name(),
-				Image:               regImage,
-				ImageIndex:          regImageIndex,
-			})
-			errCh <- nil
+	errChVerifyImages := make(chan error, len(imgOrIndexes))
+	for _, item := range imgOrIndexes {
+		item := item // copy
+
+		go func() {
+			importThrottle.Take()
+			defer importThrottle.Done()
+
+			o.verifyProcessedImages(item, errChVerifyImages, importRepo, registry, importedImages)
 		}()
 	}
 
+	err = checkForAnyAsyncErrors(imgOrIndexes, errChVerifyImages)
+	if err != nil {
+		return nil, err
+	}
+
+	return importedImages, nil
+}
+
+func checkForAnyAsyncErrors(imgOrIndexes []imagedesc.ImageOrIndex, errCh chan error) error {
 	for i := 0; i < len(imgOrIndexes); i++ {
 		err := <-errCh
 		if err != nil {
-			return nil, err
+			return err
 		}
 	}
+	return nil
+}
 
-	return importedImages, nil
+func addImageOrImageIndexToTaggableMapForMultiWrite(item imagedesc.ImageOrIndex, errCh chan error, importRepo regname.Repository, registry ImagesReaderWriter, imageOrIndexesToWrite map[regname.Reference]regremote.Taggable, lock *sync.Mutex) {
+	itemDigest, err := item.Digest()
+	if err != nil {
+		errCh <- err
+		return
+	}
+
+	tag := fmt.Sprintf("imgpkg-%s-%s", itemDigest.Algorithm, itemDigest.Hex)
+	uploadTagRef, err := regname.NewTag(fmt.Sprintf("%s:%s", importRepo.Name(), tag))
+	if err != nil {
+		errCh <- fmt.Errorf("Building upload tag image ref: %s", err)
+		return
+	}
+
+	switch {
+	case item.Image != nil:
```

Comment:
we should panic in default switch case if it's not image or index.

Reply:
I thought the saying went to stay calm and not panic? 

I'll add the panic to the case.